### PR TITLE
Enhancement: ATC dryrun

### PR DESCRIPTION
**Pull request by barnjamin**
Adding `dryrun` here to see how it feels, added cost to abi result as well but maybe should move that elsewhere?

I copied most of the body from `execute` method

If this looks reasonable I can add to other sdks

api is:
```py
        signer = AccountTransactionSigner(pk)

        sp = client.suggested_params()
        # Create atc to handle method calling for us
        atc = AtomicTransactionComposer()
        # add a method call to "call" method, pass the second app id so we can dispatch a call
        atc.add_method_call(
            app_id, get_method(c, "acct_param"), addr, sp, signer, method_args=[addr]
        )

        result = atc.dryrun(client)

        for i, res in enumerate(result.abi_results):
            print(
                "Result of '{}': {} (cost {})".format(
                    atc.method_dict[i].name, res.return_value, res.cost
                )
            )
```

Contains:
```py
dryrun_response: dict,
tx_ids: List[str],
results: DryrunABIResult,
traces: List[DryrunResponse],
```

### CODE REVIEW

**Comment by algochoi** [comment]
We have this [PR open for C2C testing](https://github.com/algorand/py-algorand-sdk/pull/268/files) that includes the `PnedingTransactionInfo` into the `ABIResult`, which might make checking for cost either. Might be relevant: https://github.com/algorand/py-algorand-sdk/blob/e1aa79c23dd5f97b67181cb88f4053d2efe12ec4/algosdk/atomic_transaction_composer.py#L694

**Comment by barnjamin** [comment]
@algochoi I think only the DryrunTxnResult contains the cost field

https://github.com/algorand/go-algorand/blob/8a335bb2b10ea3cb6e9471526701bcaa4089b53d/daemon/algod/api/server/v2/generated/types.go#L303

What do you think about abstracting the log parsing into a self contained method that gets re-used?

**Comment by algochoi** [comment]
> @algochoi I think only the DryrunTxnResult contains the cost field
> 
> https://github.com/algorand/go-algorand/blob/8a335bb2b10ea3cb6e9471526701bcaa4089b53d/daemon/algod/api/server/v2/generated/types.go#L303
> 
> What do you think about abstracting the log parsing into a self contained method that gets re-used?

Ahh okay, that makes sense. So the `ABIResult` object will only have a `cost` field for the dryrun object (or maybe we can create/inherit a separate dryrun response object)? I think the log parsing method looks good to me, but others can chime in on this.

**Review by jasonpaulos — COMMENTED** [review]
I like the idea of this, but it might be clearer to have a different response type for the `dryrun` method instead of reusing `AtomicTransactionResponse`.

**Comment by barnjamin** [comment]
@jasonpaulos thats fair, can I keep the log parsing as is?

**Comment by jasonpaulos** [comment]
> @jasonpaulos thats fair, can I keep the log parsing as is?

To clarify, I think the dryrun response should not use `AtomicTransactionResponse` or `ABIResult`. It should likely have its own response type(s) that contain only the fields it will populate.

Though I agree that the same log parsing code should be used to find the return values for both dryrun and normal execute. It just probably should not return `ABIResult`s, instead maybe a tuple of the raw information that can be used to create an `ABIResult` or the dryrun equivalent.

File: algosdk/atomic_transaction_composer.py [hunk 1]
```diff
@@ -703,3 +747,31 @@ def __init__(
         self.confirmed_round = confirmed_round
         self.tx_ids = tx_ids
         self.abi_results = results
+
+
+class DryrunABIResult:
```
Comment (algochoi) [inline algosdk/atomic_transaction_composer.py:752]: How about we inherit this from the `ABIResult` class? This constructor can call `super()` and set the cost.

**Review by algochoi — COMMENTED** [review]
Looks largely good to me, thanks for doing this. I would also suggest merging the `tzaffi/c2c` branch to here because there may be some conflicting changes (e.g. we added a `tx_info` field to the `ABIResult` class)

**Comment by barnjamin** [comment]
>I would also suggest merging the tzaffi/c2c branch to here because there may be some conflicting changes

Merged, probably should wait to review this again until that one is merged to master though?

**Comment by jasonpaulos** [comment]
@barnjamin if can merge develop back into this branch, I can give it a final review

**Comment by barnjamin** [comment]
@jasonpaulos does it make sense to wait for the dryrun stack printer to be merged in first? then the result of calling atc.dryrun also includes the DryrunRequest object?

**Comment by barnjamin** [comment]
@algochoi 
> Are we going to add this to the other SDKs and add tests like https://github.com/algorand/algorand-sdk-testing/pull/161/files

I expect it will need to be added to other SDKs as well but I think before adding tests we need to find a nice way to test the dryrun request creation bits using some live-ish ledger.

Alternatively we could just create some DryrunResponse json objects and make sure it parses out the abi results correctly?

**Comment by barnjamin** [comment]
@jasonpaulos @ahangsu can I get another peek at this please? 

Other SDKs will replicate the behavior, most of the value here is in decoding the ABI results

**Pull request closed without merging**